#163562508 Save Email Automations to the Database

janetnim opened the pull request:
#### What does this PR do?
- Fixes bug on saving automation emails to the database


#### Description of Task to be completed?
- Save email automations to the database

#### How should this be manually tested?
1.  Git clone the repository
2. Checkout to `bg-save-email-automations-163562508`
3. Install necessary dependencies
4. Ensure the `.env` file is updated with the required variables
5. Start the server `yarn start:dev' 
6. Create a new placement and access email automations using the endpoint `localhost:8000/api/v1/automations`
7. Check the `emailAutomation` table as well to see the saved email automations


#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[163562508](https://www.pivotaltracker.com/story/show/163562508)

#### Postman Documentation
N/A

#### Screenshots (If applicable)
<img width="1440" alt="screen shot 2019-02-11 at 12 46 08" src="https://user-images.githubusercontent.com/29487740/52555555-21493d80-2dfb-11e9-9d54-65b496bcaaab.png">

valmayaki commented on test/endpoints/automations.test.js:
```diff
@@ -1,15 +1,35 @@
+    console.log(result);
+    console.log('====================================');
+  });
+
```
@janetnim 
```suggestion
describe('Tests for automation endpoints\n', () => {
  var axiosStub, dateStub;
  before(async () => {
    axiosStub = sinon.stub(axios, 'get').callsFake(
      () => new Promise((resolve) => {
        resolve(automationData);
      }),
    );
    dateStub = sinon.stub(Date, 'now').callsFake(() => 1542376484108);
    const result = await resource.fetchNewPlacements('Placed - Awaiting Onboarding', 3);
    console.log('====================================');
    console.log(result);
    console.log('====================================');
  });
  after(()=> {
    axiosStub.restore();
    dateStub.restore();
  })
```
Always restore the stubbed object to prevent it from affecting other test stubs for example the error in circle ci
```bash
2) Partner and Allocations Test Suite
       Should fetch new placements based on the status specified, and the numberOfDays provided:
     TypeError: Attempted to wrap get which is already wrapped
      at checkWrappedMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:38:21)
      at wrapMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:85:13)
      at stub (node_modules/sinon/lib/sinon/stub.js:71:44)
      at Sandbox.stub (node_modules/sinon/lib/sinon/sandbox.js:291:33)
      at Context.stub (test/modules/allocations.test.js:11:32)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:62:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:288:22)
      at Generator.prototype.(anonymous function) [as next] (node_modules/regenerator-runtime/runtime.js:114:21)
      at asyncGeneratorStep (test/modules/allocations.test.js:21:103)
      at _next (test/modules/allocations.test.js:23:194)
      at /home/circleci/repo/test/modules/allocations.test.js:23:364
      at new Promise (<anonymous>)
      at Context.<anonymous> (test/modules/allocations.test.js:23:97)
  --------------
  Error: Stack Trace for original
      at wrapMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:110:31)
      at stub (node_modules/sinon/lib/sinon/stub.js:71:44)
      at Sandbox.stub (node_modules/sinon/lib/sinon/sandbox.js:291:33)
      at Context.stub (test/endpoints/automations.test.js:13:11)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:62:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:288:22)
      at Generator.prototype.(anonymous function) [as next] (node_modules/regenerator-runtime/runtime.js:114:21)
      at asyncGeneratorStep (test/endpoints/automations.test.js:21:103)
      at _next (test/endpoints/automations.test.js:23:194)
      at /home/circleci/repo/test/endpoints/automations.test.js:23:364
      at new Promise (<anonymous>)
      at Context.<anonymous> (test/endpoints/automations.test.js:23:97)

  3) Partner and Allocations Test Suite
       Should throw an error when no partner record was found:
     TypeError: Attempted to wrap get which is already wrapped
      at checkWrappedMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:38:21)
      at wrapMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:85:13)
      at stub (node_modules/sinon/lib/sinon/stub.js:71:44)
      at Sandbox.stub (node_modules/sinon/lib/sinon/sandbox.js:291:33)
      at Context.stub (test/modules/allocations.test.js:57:32)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:62:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:288:22)
      at Generator.prototype.(anonymous function) [as next] (node_modules/regenerator-runtime/runtime.js:114:21)
      at asyncGeneratorStep (test/modules/allocations.test.js:21:103)
      at _next (test/modules/allocations.test.js:23:194)
      at /home/circleci/repo/test/modules/allocations.test.js:23:364
      at new Promise (<anonymous>)
      at Context.<anonymous> (test/modules/allocations.test.js:23:97)
  --------------
  Error: Stack Trace for original
      at wrapMethod (node_modules/sinon/lib/sinon/util/core/wrap-method.js:110:31)
      at stub (node_modules/sinon/lib/sinon/stub.js:71:44)
      at Sandbox.stub (node_modules/sinon/lib/sinon/sandbox.js:291:33)
      at Context.stub (test/endpoints/automations.test.js:13:11)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:62:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:288:22)
      at Generator.prototype.(anonymous function) [as next] (node_modules/regenerator-runtime/runtime.js:114:21)
      at asyncGeneratorStep (test/endpoints/automations.test.js:21:103)
      at _next (test/endpoints/automations.test.js:23:194)
      at /home/circleci/repo/test/endpoints/automations.test.js:23:364
      at new Promise (<anonymous>)
      at Context.<anonymous> (test/endpoints/automations.test.js:23:97)

```

janetnim replied:
Thanks for that. I'll implement the changes suggested

valmayaki commented on test/endpoints/automations.test.js:
```diff
@@ -1,15 +1,35 @@
+    console.log('====================================');
+  });
+
+  it('Should return automation data with a 200 response code', async (done) => {
```
@janetnim  don't use async and done together use either one of them
```suggestion
  it('Should return automation data with a 200 response code', (done) => {
```
this will eliminate the error you where getting in circle ci
```bash
 Should return automation data with a 200 response code:
     Error: Resolution method is overspecified. Specify a callback *or* return a Promise; not both.
      at done (test/endpoints/automations.test.js:40:9)
      at Test.Request.callback (node_modules/superagent/lib/node/index.js:716:12)
      at parser (node_modules/superagent/lib/node/index.js:916:18)
      at IncomingMessage.res.on (node_modules/superagent/lib/node/parsers/json.js:19:7)
      at endReadableNT (_stream_readable.js:1081:12)
      at process._tickCallback (internal/process/next_tick.js:63:
```

janetnim replied:
I've made the said changes on using async and done at the same time

nzediegwu1 commented on server/jobs/offboarding/email.js:
```diff
@@ -9,6 +10,9 @@ import { executeEmailAutomation } from '../helpers';
-  executeEmailAutomation([sendSOPOffboardingMail, sendITOffboardingMail], placement);
-}
+const emailOffboarding = async (placement) => {
+  placement = await fetchNewPlacements('External Engagements - Rolling Off', 1);
```
Nice job. However I think `fetchNewPlacements` function no longer takes two arguments from the fix being done in replacing mock-api post endpoints with seeder

nzediegwu1 replied:
The `placement` variable already contains the data this job needs in its parameter, hence no need calling `fetchNewPlacements` again

nzediegwu1 commented on server/jobs/onboarding/email.js:
```diff
@@ -1,16 +1,18 @@
 const emailOnboarding = async (placement) => {
-  executeEmailAutomation([sendDevOnboardingMail, sendSOPOnboardingMail], placement);
+  placement = await fetchNewPlacements('Placed - Awaiting Onboarding', 1);
+  executeEmailAutomation([sendDevOnboardingMail, sendSOPOnboardingMail], placement[0]);
```
Same here

nzediegwu1 commented on test/endpoints/automations.test.js:
```diff
@@ -1,15 +1,36 @@
       .request(app)
       .get('/api/v1/automations')
       .end((err, res) => {
+        console.log('=============BODYYY=======================');
```
Nice to have written tests. Please remove these `console.log`s

nzediegwu1 commented on test/helpers/helpers.test.js:
```diff
@@ -23,7 +23,8 @@ describe('Test that helper functions work as expected', () => {
     expect(mailInfo.partnerName).to.equal(placement.client_name);
-    expect(mailInfo.partnerLocation).to.equal(samplePartner.data.values[0].location);
+    // expect(mailInfo.partnerLocation).to.equal(samplePartner.data.values[0].location);
+    // above will be uncommented when the bug on fetching partner details has been worked on
```
Noted

janetnim commented:
@nzediegwu1 I made the changes requested

nzediegwu1 commented on server/index.js:
```diff
@@ -31,7 +31,7 @@ app.get('*', (req, res) => res.status(200).send({
 }));
 
-const port = parseInt(process.env.PORT, 10) || 8000;
+const port = parseInt(process.env.PORT, 10) || 6050;
```
What is the reason for this change in `port`

nzediegwu1 commented on server/jobs/helpers.js:
```diff
@@ -16,7 +16,10 @@ export const getMailInfo = async (placement) => {
   } = placement;
-  const { location: partnerLocation } = await findPartnerById(partnerId);
+
+  // const { location: partnerLocation } = await findPartnerById(partnerId);
```
Is this not supposed to be fixed in this PR?

janetnim replied:
It was previously another story but am looking to solve this in this PR

nzediegwu1 commented:
From the file changes, I couldn't see where the implementation for `saving email automation to database` as this PR suggests. Did you mistakenly give the PR a wrong description?

janetnim commented:
I was to fix the bug on why emails were not being saved into the DB but all the fixes I made before were replaced when your PR eas merged and only the partner location was the problem which I later tested and proved to work eventually. I just modified some test instead and added time interval env variable to codeclimate.
